src/Ledger/Conway/PoolReap.lagda.md

```diff
@@ -0,0 +1,110 @@
+# Pool Reaping Transition {sec:pool-reaping-transition}
+
+<!--
+```agda
+{-# OPTIONS --safe #-}
+
+open import Ledger.Conway.Abstract
+open import Ledger.Conway.Transaction
+
+module Ledger.Conway.PoolReap
+  (txs : _) (open TransactionStructure txs)
+  (abs : AbstractFunctions txs)
+  where
+open import Ledger.Prelude
+open import Ledger.Conway.Utxo txs abs
+open import Ledger.Conway.Certs govStructure
+```
+-->
+
+```agda
+record PoolReapState : Type where
+```
+<!--
+```agda
+  constructor ⟦_,_,_,_⟧ᵖ
+```
+```agda
+  field
+    utxoSt     : UTxOState   -- utxo state
+    acnt       : Acnt        -- accounting
+    dState     : DState      -- delegation state
+    pState     : PState      -- pool state
+```
+
+<!--
+```agda
+instance
+  unquoteDecl HasCast-PoolReapState = derive-HasCast
+                [ (quote PoolReapState , HasCast-PoolReapState) ]
+```
+
+Recall, `PState`{.AgdaDatatype} is a record with two fields, `pools`{.AgdaField}
+and `retiring`{.AgdaField} (maps on `KeyHash`{.AgdaField} with codomains
+`PoolParams`{.AgdaRecord} and `Epoch`{.AgdaDatatype}, respectively).
+`PoolParams`{.AgdaRecord} is a record with just one field, the
+`rewardAddr`{.AgdaField} credential.
+
+```agda
+private variable
+  e lastEpoch : Epoch
+  poolReapState : PoolReapState
+  pp : PParams
+
+data _⊢_⇀⦇_,POOLREAP⦈_ : PParams → PoolReapState → Epoch → PoolReapState → Type where
+  POOLREAP : let
+    -- open LState ls
+    open PoolReapState poolReapState
+    open PoolParams
+    open UTxOState
+    open PState
+    open DState
+    open Acnt
+    open PParams
+
+    retired    = (pState .retiring) ⁻¹ e
+    pr = constMap retired  (pp .poolDeposit)
+    rewardAcnts = mapValues rewardAccount $ (pState .pools) ∣ retired
+
+--    rewardAcnts' : RwdAddr ⇀ Coin
+--    rewardAcnts' =?= constMap (range (rewardAcnts ˢ)) (pp .poolDeposit)
+
+    rewardsAcnts' =
+      let raccounts = range rewardAcnts
+          combineDeposits : Credential → ℕ
+          combineDeposits a = foldr _+_ 0 (rewardAcnts ⁻¹ a)
+          retiredWithRewardAccounts =
+            mapˢ (λ y → (y , combineDeposits y)) raccounts
+       in retiredWithRewardAccounts , left-unique-mapˢ raccounts
```

Comment:
Here is my last roadblock. I'm trying to define a map from reward accounts to the amount to be reimbursed.

When more than one pool uses the same reward address, we are supposed to add up the deposits corresponding to each pool.

At the moment, the type checking error I get from Agda is that I cannot add the elements of a set of Coin with `foldr`. But sets are modeled with lists as far as I can tell.

I'm missing something still.

Reply:
Fixed.